core/iwasm/common/wasm_memory.c

```diff
@@ -39,7 +39,7 @@ wasm_memory_init_with_pool(void *mem, unsigned int bytes)
     mem_allocator_t _allocator = mem_allocator_create(mem, bytes);
 
     if (_allocator) {
-        memory_mode = MEMORY_MODE_POOL;
+        __atomic_store_n(&memory_mode, MEMORY_MODE_POOL, __ATOMIC_SEQ_CST);
```

Comment:
I remember that the global variable `memory_mode` is set only when wasm_runtime_init/wasm_runtime_full_init, and won't be set again after that. Is there any data race risk on it?

Reply:
this is probably the only exception (done when destroying the memory)
https://github.com/bytecodealliance/wasm-micro-runtime/blob/739acfc9082e63967ee91fca85dc01ab1630a5cc/core/iwasm/common/wasm_memory.c#L131
and it's causing the data race

Reply:
The root cause of `memory_mode` data race is not this file, it is due to main thread doesn't wait the destroying of sub thread's exec env, and start to destroy the runtime environment (only when all child threads exit and all exec envs are destroyed can the wasm_runtime_destroy be called). And then when sub thread's exec env is destroyed, it reads an re-initialized `memory_mode`.
I uploaded PR #1973 to fix it, could you help try?

Reply:
Yes, that was my guess too, but I didn't have time to try it. I'll check the change and let you know in #1973